vortex-file/src/read/mask.rs

```diff
@@ -1,98 +1,121 @@
 use std::cmp::{max, min};
 use std::fmt::{Display, Formatter};
 
-use arrow_buffer::{BooleanBuffer, MutableBuffer};
-use croaring::Bitmap;
+use arrow_buffer::{bit_util, BooleanBuffer, BooleanBufferBuilder, MutableBuffer};
 use vortex_array::array::{BoolArray, PrimitiveArray, SparseArray};
-use vortex_array::compute::{filter, slice, take, TakeOptions};
+use vortex_array::compute::{and, filter, slice, take, TakeOptions};
+use vortex_array::stats::ArrayStatistics;
 use vortex_array::validity::{LogicalValidity, Validity};
-use vortex_array::{iterate_integer_array, ArrayData, IntoArrayData, IntoArrayVariant};
-use vortex_dtype::PType;
-use vortex_error::{vortex_bail, vortex_err, VortexResult};
-
+use vortex_array::{iterate_integer_array, ArrayDType, ArrayData, IntoArrayData, IntoArrayVariant};
+use vortex_dtype::{DType, Nullability, PType};
+use vortex_error::{vortex_bail, vortex_err, VortexExpect, VortexResult, VortexUnwrap};
 const PREFER_TAKE_TO_FILTER_DENSITY: f64 = 1.0 / 1024.0;
 
 /// Bitmap of selected rows within given [begin, end) row range
-#[derive(Debug, Clone, Default, PartialEq, Eq)]
+#[derive(Debug, Clone)]
 pub struct RowMask {
-    values: Bitmap,
+    bitmask: ArrayData,
     begin: usize,
     end: usize,
 }
 
+impl PartialEq for RowMask {
+    fn eq(&self, other: &Self) -> bool {
+        self.begin == other.begin
+            && self.end == other.end
+            && self
+                .bitmask
+                .clone()
+                .into_bool()
```

Comment:
this is really expensive (going to canonicalize and then throw away) if the array isn't a BoolArray

Reply:
yes, this is only for tests. I can make this impl only exist in test mode